File: src/poincareview.cpp
#include "poincareview.h"
#include "reflectionaxis.h"
#include "math_helpers.h"

#include <QWidget>
#include <QPainter>
#include <QPainterPath>
#include <QVector>
#include <QBrush>
#include <QRegion>
#include <QString>
#include <QOpenGLFunctions>

#include <algorithm>
#include <iostream>
#include <math.h>
#include <vector>
#include <unordered_set>

PoincareView::PoincareView(QWidget *parent) :
	QOpenGLWidget(parent) {

	this->centerVertices = new QVector<QPointF *>();
	this->origin = new QPointF();

	this->drawnCount = 0;
	this->drawnTiles = std::map<float, std::unordered_set<float> >();
	this->sideCount = 7;
	this->adjacentCount = 3;
	this->renderLayers = 6;
}
PoincareView::~PoincareView(){}

void PoincareView::genCenterVertices() {
	int p = sideCount;
	int q = adjacentCount;
	//Put this line in its own method
	float dist = (diskDiameter/2) * sqrt(cos(M_PI/p + M_PI/q)*cos(M_PI/q) / (sin(2*M_PI/q) * sin(M_PI/p) + cos(M_PI/p + M_PI/q)* cos(M_PI/q)));
	float alpha = 2 * M_PI / sideCount;

	for(u_int i = 0; i < sideCount; i++) {
		float x = origin->x() + (dist) * cos(i * alpha);
		float y = origin->y() + (dist) * sin(i * alpha);

		centerVertices->push_back(new QPointF(x, y));
	}
}

bool PoincareView::hasBeenDrawn(QPointF *aPoint) {
	int precision = pow(10,3);
	float x = round(precision * aPoint->x())/precision;
	float y = round(precision * aPoint->y())/precision;

	if(drawnTiles.count(x) == 1) {
		if(drawnTiles[x].count(y) == 1)
			return true;
	} else {
		drawnTiles[x] = std::unordered_set<float>();
	}

	drawnTiles[x].insert(y);
	return false;
}

void PoincareView::drawTile(QVector<QPointF *> *vertices, QPointF *center, int layers) {
	QVector<QPointF *> *reflectedVertices = new QVector<QPointF *>(sideCount);
	QPointF *reflectedCenter;
	ReflectionAxis *axis;

	if(hasBeenDrawn(center))
		return;

	drawnCount++;
	painter->drawPoint(*center);

	//Draw and reflect across each side of this tile
	for(auto it = vertices->begin(); it != vertices->end(); ++it) {
		QPointF *A = *it;
		QPointF *B = (it + 1 != vertices->end())
			? *(it + 1)
			: *(vertices->begin());
	
		if(areCollinear(A, B, origin))
			axis = new LineAxis(A, B);
		else
			axis = new ArcAxis(A, B, origin, diskDiameter);

		axis->draw(painter);
		if(layers == 1)
			continue;

		reflectedCenter = axis->reflectPoint(center);
		reflectedVertices->clear();

		for(auto j : *vertices) {
			reflectedVertices->push_back(axis->reflectPoint(j));
		}

		drawTile(reflectedVertices, reflectedCenter, layers - 1);
	}
}

void PoincareView::paintEvent(QPaintEvent *e) {
	if(e){}
	this->diskDiameter = (int)std::min(size().width(), size().height()) - 10;
	this->painter = new QPainter(this);
	painter->setRenderHint(QPainter::HighQualityAntialiasing);
	this->diskPath = new QPainterPath();

	this->origin->setX(round(size().width()/2));
	this->origin->setY(round(size().height()/2));

	painter->setPen(QPen(QColor(122, 0, 127, 255), 2));
	this->diskRegion = new QRegion(QRect(origin->x() - diskDiameter/2, origin->y() - diskDiameter/2, diskDiameter, diskDiameter), QRegion::Ellipse);
	diskPath->addRegion(*diskRegion);
	painter->setClipRegion(*diskRegion);

	genCenterVertices();
	drawTile(this->centerVertices, this->origin, this->renderLayers);

	painter->setClipping(false);

	int diskCenterX = -diskDiameter/2 + origin->x();
	int diskCenterY = -diskDiameter/2 + origin->y();
	painter->setPen(QPen(QColor(5, 0, 127, 255), 3));
	painter->drawEllipse(QRect(diskCenterX, diskCenterY, diskDiameter, diskDiameter));

	painter->end();
	std::cout << drawnCount << std::endl;

	delete this->diskPath;
	this->drawnCount = 0;
	this->drawnTiles.clear();
	this->centerVertices->clear();
}

File: src/poincareview.py
from PyQt5.QtGui import *
from PyQt5.QtWidgets import *
from PyQt5.QtCore import *

from reflectionaxis import ArcAxis, LineAxis
from math_helpers import areCollinear

import math
from collections import defaultdict
from pprint import *

class PoincareView(QOpenGLWidget):
	def __init__(self):
		super().__init__()
		self.centerVertices = []
		self.origin = QPointF()

		self.drawnCount = 0
		self.drawnTiles = defaultdict(set)
		self.sideCount = 5
		self.adjacentCount = 4
		self.renderLayers = 3

	def genCenterVertices(self):
		p = self.sideCount
		q = self.adjacentCount
		dist = (self.diskDiameter/2) * math.sqrt(math.cos(math.pi/p + math.pi/q)*math.cos(math.pi/q) / (math.sin(2*math.pi/q) * math.sin(math.pi/p) + math.cos(math.pi/p + math.pi/q)* math.cos(math.pi/q)))
		alpha = 2*math.pi/p

		for i in range(p):
			x = self.origin.x() + (dist) * math.cos(i * alpha)
			y = self.origin.y() + (dist) * math.sin(i * alpha)

			self.centerVertices.append(QPointF(x, y))
		
	def hasBeenDrawn(self, aPoint):
		precision = 1000
		x = round(precision * aPoint.x())/precision
		y = round(precision * aPoint.y())/precision

		if x in self.drawnTiles and y in self.drawnTiles[x]:
			return True

		self.drawnTiles[x].add(y)
		return False

	def drawTile(self, vertices, center, layers) :
		reflectedVertices = []
		reflectedCenter = QPointF()

		self.drawnCount += 1
		self.painter.drawPoint(center)

		for i, v in enumerate(vertices):
			A = v
			if i+1 is not len(vertices):
				B = vertices[i+1]
			else:
				B = vertices[0]

			if areCollinear(A, B, self.origin):
				self.painter.drawLine(A, B)
				continue
			else:
				axis = ArcAxis(A, B, self.origin, self.diskDiameter)
				if axis.collinear:
					axis = LineAxis(A, B)

		
			axis.draw(self.painter)

			if layers is 1: continue

			reflectedCenter = axis.reflectPoint(center)
			if self.hasBeenDrawn(reflectedCenter):
				continue
			reflectedVertices = [axis.reflectPoint(j) for j in vertices]

			self.drawTile(reflectedVertices, reflectedCenter, layers - 1) == 0
		
	def paintEvent(self, QPaintEvent):
		self.diskDiameter = min(self.size().width(), self.size().height()) - 10
		self.painter = QPainter(self)
		self.diskPath = QPainterPath()

		self.origin.setX(self.size().width()/2)
		self.origin.setY(self.size().height()/2)

		self.painter.setPen(QPen(QColor(122, 0, 127, 255), 2))

		diskRegion = QRegion(QRect(self.origin.x() - self.diskDiameter/2, self.origin.y() - self.diskDiameter/2, self.diskDiameter, self.diskDiameter), QRegion.Ellipse)
		self.painter.setClipRegion(diskRegion)

		self.genCenterVertices()
		self.painter.drawPoint(self.origin)
		self.drawTile(self.centerVertices, self.origin, self.renderLayers)

		self.painter.setClipping(False)

		diskCenterX = -1 * self.diskDiameter/2 + self.origin.x()
		diskCenterY = -1 * self.diskDiameter/2 + self.origin.y()
		self.painter.setPen(QPen(QColor(5, 0, 127, 255), 3))
		self.painter.drawEllipse(QRect(diskCenterX, diskCenterY, self.diskDiameter, self.diskDiameter))

		self.painter.end()
		del self.painter
		self.centerVertices = []
		print(self.drawnCount)
		self.drawnCount = 0
		self.drawnTiles = defaultdict(set)
